core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java

```diff
@@ -86,32 +86,30 @@ public void onNewClusterState(ClusterState state) {
 
             @Override
             public void onClusterServiceClose() {
-                logger.warn("[{}] cluster service closed while waiting for write consistency number of shards " +
-                                "to be allocated on index creation.", indexName);
-                listener.onResponse(new ClusterStateUpdateResponse(false));
+                logger.debug("[{}] cluster service closed while waiting for enough shards to be started.", indexName);
+                listener.onFailure(new NodeClosedException(clusterService.localNode()));
             }
 
             @Override
             public void onTimeout(TimeValue timeout) {
-                logger.warn("[{}] index was created, but the operation timed out waiting for write consistency " +
-                                "number of shards to be allocated.", indexName);
-                listener.onResponse(new ClusterStateUpdateResponse(false));
+                logger.debug("[{}] index created, but the operation timed out while waiting for enough shards to be started.", indexName);
+                listener.onFailure(new ElasticsearchTimeoutException("[{}] index created, but the operation timed out while " +
+                                                                     "waiting for enough shards to be started.", indexName));
             }
         };
 
-        if (observedState.status() == ClusterState.ClusterStateStatus.APPLIED
-                && enoughShardsActive(observedState, indexName, settings, waitOnActiveShards)) {
-            observerListener.onNewClusterState(observedState);
-        } else {
-            observer.waitForNextChange(observerListener, shardsAllocatedPredicate, timeout);
-        }
+        observer.waitForNextChange(observerListener, shardsAllocatedPredicate, timeout);
     }
 
     // package private for testing
     static boolean enoughShardsActive(final ClusterState clusterState, final String indexName,
-                                      final Settings settings, int waitOnActiveShards) {
-        if (waitOnActiveShards == ReplicationOperation.WAIT_ON_ACTIVE_SHARDS_NOT_SET) {
-            waitOnActiveShards = WAIT_ON_ACTIVE_SHARDS_SETTING.get(settings);
+                                      final Settings settings, ActiveShardCount waitForActiveShards) {
+        if (waitForActiveShards.value() == 0) {
+            // not waiting for any active shards
+            return true;
+        }
+        if (waitForActiveShards == ActiveShardCount.DEFAULT) {
+            waitForActiveShards = WAIT_FOR_ACTIVE_SHARDS_SETTING.get(settings);
         }
         final IndexRoutingTable indexRoutingTable = clusterState.routingTable().index(indexName);
         assert indexRoutingTable != null;
```

Comment:
what I wrote here is not true. The master node itself only acks once it has applied the state locally.